src/PublicAllocator.sol

```diff
@@ -1,25 +1,127 @@
+
+import {Market} from "../lib/metamorpho/lib/morpho-blue/src/interfaces/IMorpho.sol";
+import {UtilsLib} from "./libraries/UtilsLib.sol";
+
+import {ErrorsLib} from "./libraries/ErrorsLib.sol";
+import {EventsLib} from "./libraries/EventsLib.sol";
+import {FlowCap, FlowConfig, SupplyConfig, IPublicAllocatorStaticTyping} from "./interfaces/IPublicAllocator.sol";
+
+contract PublicAllocator is IPublicAllocatorStaticTyping {
+    using MorphoLib for IMorpho;
+    using MorphoBalancesLib for IMorpho;
+    using MarketParamsLib for MarketParams;
+    using SharesMathLib for uint256;
+    using UtilsLib for uint256;
+    using UtilsLib for uint128;
+
+    /// CONSTANTS ///
+
+    address public immutable OWNER;
+    IMorpho public immutable MORPHO;
+    IMetaMorpho public immutable VAULT;
 
     /// STORAGE ///
 
-    address public owner;
-    IMetaMorpho public vault;
+    uint256 public fee;
+    mapping(Id => FlowCap) public flowCap;
+    mapping(Id => uint256) public supplyCap;
+
+    /// MODIFIER ///
+
+    modifier onlyOwner() {
+        if (msg.sender != OWNER) revert ErrorsLib.NotOwner();
+        _;
+    }
 
     /// CONSTRUCTOR ///
 
-    constructor(address newOwner, IMetaMorpho newVault) {
-        owner = newOwner;
-        vault = newVault;
+    constructor(address newOwner, address vault) {
+        if (newOwner == address(0)) revert ErrorsLib.ZeroAddress();
+        if (vault == address(0)) revert ErrorsLib.ZeroAddress();
+        OWNER = newOwner;
+        VAULT = IMetaMorpho(vault);
+        MORPHO = VAULT.MORPHO();
     }
 
     /// PUBLIC ///
 
-    function reallocate(MarketAllocation[] calldata allocations) external {
-        vault.reallocate(allocations);
+    function reallocate(MarketAllocation[] calldata allocations) external payable {
+        if (msg.value != fee) revert ErrorsLib.IncorrectFee(msg.value);
+
+        uint256[] memory assets = new uint256[](allocations.length);
+        for (uint256 i = 0; i < allocations.length; ++i) {
+            // Do not compute interest twice for every market
+            MORPHO.accrueInterest(allocations[i].marketParams);
+            assets[i] = MORPHO.expectedSupplyAssets(allocations[i].marketParams, address(VAULT));
+        }
+
+        VAULT.reallocate(allocations);
+
+        MarketParams memory marketParams;
+        for (uint256 i = 0; i < allocations.length; ++i) {
+            marketParams = allocations[i].marketParams;
+            Id id = marketParams.id();
+            uint256 newAssets = MORPHO.expectedSupplyAssets(marketParams, address(VAULT));
+            if (newAssets >= assets[i]) {
+                if (newAssets > supplyCap[id]) revert ErrorsLib.PublicAllocatorSupplyCapExceeded(id);
+                uint128 inflow = (newAssets - assets[i]).toUint128();
+                flowCap[id].maxIn -= inflow;
+                flowCap[id].maxOut = (flowCap[id].maxOut).saturatingAdd(inflow);
```

Comment:
if you put twice the same market in `allocations`, `maxIn` and `maxOut` are updated twice no? It seems that you can game the publicAllocator this way.

Reply:
Damn yes, you can make the public allocator overestimate the flows.